### PR TITLE
EIP support for the VM and Common

### PR DESCRIPTION
**Pull request by holgerd77**
This is my suggestion for a gentle introduction on EIP support for the VM. The solution plays well with our current setup, is backwards-compatible but should also be flexible enough to carry us on the first introductions here (and maybe will turn out sufficient as a longer-term solution as well).

So this can be used in the VM as follows:

```typescript
if (this._activatedEIPs.includes('EIP2537') {
  const gasPrice = this._common.paramByEIP('gasPrices', 'Bls12381G1AddGas', 'EIP2537')
  // Activate the precompiles or whatever is to be done
}
```

Once EIPs move into HFs we can easily shift the EIP config in `Common` to the respective HF config on the proposed solution.

Some design decisions:

1. **EIP names for API on the VM**: I've chosen the string version here - so e.g. `'EIP2537'` and not the plain number `2537`. I find this better readable since EIPs are always stated with an EIP-prefix and I don't see a downside. But at the end this is a bit a matter of subjective taste and I would also be open for a change if someone has strong opinions on this.

2. **param() vs paramByEIP()**: `paramByEIP()` could theoretically also have been integrated into `param()` and the function would still be well-defined. I decided it's not worth it and overloads the semantics of this function with the need for too much textual explanation, e.g. on the case of (as it is now) automatically falling back to the HF-centric version when no HF is provided. So - therefore - an extra function, think this doesn't hurt.

3. **EIP and HF dependencies**: I couldn't decide on a design decision for eventual EIP and/or HF dependencies (so if e.g. an EIP depends on another EIP or should only be executed in a certain HF environment) - as we talked about in the call. I especially couldn't decide if to integrate in Common or the VM directly. So for now I just left this, I think we can let this sink in, gain some experiences on EIP implementations and decide/add later. For EIP2537 I imagine this doesn't play a role (@jochem-brouwer: correct me if I am wrong here), since there is no EIP dependency and the EIP is so isolated in its functionality that the precompiles should run in whatever HF environment.

I've already done the `EIP2537` integration here, since this made also the test creation easier (possible).

@jochem-brouwer this might be a good starting ground for the suggestion to start fresh on #785 as suggested in the thread there (so to manually take things over) once/in case we merge here. Hope this should then be doable without too much hazzle. 😬  😄

### CODE REVIEW

File: packages/common/src/index.ts
Comment (jochem-brouwer) [inline packages/common/src/index.ts:204]: The parameter here should be `eip`, not `hardfork`

File: packages/common/src/index.ts
Comment (jochem-brouwer) [inline packages/common/src/index.ts:211]: Nit: unnecessary variable declaration

File: packages/common/tests/params.ts
Comment (jochem-brouwer) [inline packages/common/tests/params.ts:106]: Ah! Great that CodeCov reports that line 217 is not being ran - this function is not tested. Don't think I would have noticed otherwise that this function was not being called!

**Review by jochem-brouwer — CHANGES_REQUESTED** [review]
Some small nitpick comments and one test which is not being ran. Looks good in general, complete with tests and docs! Once this is merged, will merge this in the BLS PR and start using it 😄

**Comment by ryanio** [comment]
Looks great!

My only thought about **EIP names for API on the VM** is that I just converted the opcodes tables/list to Map for speed/performance reasons (#852), and I think we should use the same pattern here like `const EIPs: Map<number, EIP>`

File: packages/common/tests/params.ts
Comment (ryanio) [reply packages/common/tests/params.ts:106]: @jochem-brouwer oh that is cool, this must be new! i haven't seen this "check warning" feature before.

Screen Shot 2020-09-03 at 8 47 02 AM

File: packages/common/src/index.ts
Comment (holgerd77) [reply packages/common/src/index.ts:204]: Done (see on next push)

File: packages/common/src/index.ts
Comment (holgerd77) [reply packages/common/src/index.ts:211]: Done (thanks, please really point these things out, since these things are just unnecessarily complicating the code base)

File: packages/common/tests/params.ts
Comment (holgerd77) [reply packages/common/tests/params.ts:106]: @jochem-brouwer great catch, updated! 😄  (I am pretty sure that I wrote at least the corresponding message at some point, not sure where this got lost again? 🤔  Anyhow.)

**Comment by holgerd77** [comment]
@ryanio had a short look but I can't really imagine what should mapped to what respectively how that would look like.  Another aspect: is this really meaningful for performance on such short lists?

For now I'll leave this to @jochem-brouwer to review and eventually merge. Maybe you can do a short follow-up PR (or do along some other work) on this if it turns out we want to change.

**Review by jochem-brouwer — APPROVED** [review]

**Comment by jochem-brouwer** [comment]
I agree with @holgerd77 here, I don't think it makes much sense to use a `Map` just for performance. But if it makes our code more consistent and would not complicate things as much it might be worth a PR at some point @ryanio.

**Pull request merged**
